File: server/internal/service/gradingagent/workflow_normalize.go
package gradingagent

import (
	"log/slog"
	"strings"
)

// NormalizeWorkflowGraph rewrites legacy node types and handles to their canonical forms.
// It is idempotent and safe to call on already-normalized graphs.
func NormalizeWorkflowGraph(g *WorkflowGraph) (WorkflowGraph, int) {
	if g == nil {
		return WorkflowGraph{}, 0
	}
	changes := 0
	out := *g
	out.Nodes = make([]WorkflowNode, len(g.Nodes))
	for i, n := range g.Nodes {
		node := n
		switch node.Type {
		case NodeTypeSubmission:
			node.Type = NodeTypeStudentSubmission
			changes++
		case NodeTypeAssignmentCtx:
			node.Type = NodeTypeActivity
			changes++
		}
		out.Nodes[i] = node
	}
	nodeByID := make(map[string]WorkflowNode, len(out.Nodes))
	for _, n := range out.Nodes {
		nodeByID[n.ID] = n
	}
	normalizedEdges, edgeChanges := normalizeWorkflowEdges(g.Edges, nodeByID)
	changes += edgeChanges
	out.Edges = normalizedEdges
	if changes > 0 {
		slog.Info("grading agent workflow graph normalized", "changes", changes)
	}
	return out, changes
}

func normalizeWorkflowEdges(edges []WorkflowEdge, nodeByID map[string]WorkflowNode) ([]WorkflowEdge, int) {
	if len(edges) == 0 {
		return nil, 0
	}
	changes := 0
	out := make([]WorkflowEdge, 0, len(edges))
	for _, e := range edges {
		src, srcOK := nodeByID[e.Source]
		tgt, tgtOK := nodeByID[e.Target]
		if !srcOK || !tgtOK {
			out = append(out, e)
			continue
		}
		targetHandle := strings.TrimSpace(e.TargetHandle)

		if targetHandle == HandleContext && isPromptConsumerNodeType(tgt.Type) {
			expanded, expandedChanges := expandContextEdge(e, src, tgt.Type)
			out = append(out, expanded...)
			changes += expandedChanges
			continue
		}

		edge := e
		if tgt.Type == NodeTypeAI && isPromptConsumerLegacyTargetHandle(targetHandle) {
			edge.TargetHandle = HandleAIInput
			changes++
		}
		out = append(out, edge)
	}
	return out, changes
}

func isPromptConsumerNodeType(nodeType string) bool {
	return nodeType == NodeTypeGrader || nodeType == NodeTypeCriterionGrader || nodeType == NodeTypeAI
}

func isPromptConsumerLegacyTargetHandle(targetHandle string) bool {
	switch targetHandle {
	case HandleSubmission, HandleContent, HandleRubric, HandleContext:
		return true
	default:
		return false
	}
}

func expandContextEdge(e WorkflowEdge, src WorkflowNode, targetNodeType string) ([]WorkflowEdge, int) {
	includeContent, includeRubric := activityContextIncludeFlags(src)
	if !includeContent && !includeRubric {
		return nil, 1
	}
	changes := 1
	out := make([]WorkflowEdge, 0, 2)
	if includeContent {
		out = append(out, WorkflowEdge{
			ID:           e.ID + "-content",
			Source:       e.Source,
			SourceHandle: HandleContent,
			Target:       e.Target,
			TargetHandle: contentTargetHandleForNode(targetNodeType),
		})
	}
	if includeRubric {
		out = append(out, WorkflowEdge{
			ID:           e.ID + "-rubric",
			Source:       e.Source,
			SourceHandle: HandleRubric,
			Target:       e.Target,
			TargetHandle: rubricTargetHandleForNode(targetNodeType),
		})
	}
	return out, changes
}

func contentTargetHandleForNode(nodeType string) string {
	if nodeType == NodeTypeAI {
		return HandleAIInput
	}
	return HandleContent
}

func rubricTargetHandleForNode(nodeType string) string {
	if nodeType == NodeTypeAI {
		return HandleAIInput
	}
	return HandleRubric
}

func activityContextIncludeFlags(src WorkflowNode) (includeContent, includeRubric bool) {
	if src.Type != NodeTypeActivity {
		return false, false
	}
	if _, hadIncludeContent := src.Data["includeContent"]; hadIncludeContent {
		return boolData(src, "includeContent"), boolData(src, "includeRubric")
	}
	return true, true
}

File: clients/web/src/components/annotation/grader-agent/workflow-normalize.ts
import type { GraderWorkflowGraph } from './types'
import {
  HANDLE_AI_INPUT,
  HANDLE_CONTENT,
  HANDLE_RUBRIC,
  HANDLE_CONTEXT,
  HANDLE_SUBMISSION,
  WORKFLOW_VERSION,
} from './types'

function boolData(data: Record<string, unknown>, key: string): boolean {
  const value = data[key]
  return typeof value === 'boolean' && value
}

function activityContextIncludeFlags(data: Record<string, unknown>): {
  includeContent: boolean
  includeRubric: boolean
} {
  if ('includeContent' in data) {
    return {
      includeContent: boolData(data, 'includeContent'),
      includeRubric: boolData(data, 'includeRubric'),
    }
  }
  return { includeContent: true, includeRubric: true }
}

function isPromptConsumerNodeType(type: string): boolean {
  return type === 'grader' || type === 'criterionGrader' || type === 'ai'
}

function contentTargetHandleForNode(nodeType: string): string {
  return nodeType === 'ai' ? HANDLE_AI_INPUT : HANDLE_CONTENT
}

function rubricTargetHandleForNode(nodeType: string): string {
  return nodeType === 'ai' ? HANDLE_AI_INPUT : HANDLE_RUBRIC
}

/** Rewrites legacy node types and handles to canonical forms (mirrors server normalizer). */
export function normalizeLegacyWorkflowGraph(graph: GraderWorkflowGraph): {
  graph: GraderWorkflowGraph
  changes: number
} {
  let changes = 0
  const nodes = graph.nodes.map((node) => {
    switch (node.type) {
      case 'submission':
        changes++
        return { ...node, type: 'studentSubmission' as const }
      case 'assignmentContext':
        changes++
        return { ...node, type: 'activity' as const }
      default:
        return node
    }
  })
  const nodeById = new Map(nodes.map((node) => [node.id, node]))
  const edges: GraderWorkflowGraph['edges'] = []

  for (const edge of graph.edges) {
    const src = nodeById.get(edge.source)
    const tgt = nodeById.get(edge.target)
    if (!src || !tgt) {
      edges.push(edge)
      continue
    }
    const targetHandle = edge.targetHandle ?? ''
    if (targetHandle === HANDLE_CONTEXT && isPromptConsumerNodeType(tgt.type)) {
      const { includeContent, includeRubric } = activityContextIncludeFlags(src.data)
      changes++
      if (includeContent) {
        edges.push({
          id: `${edge.id}-content`,
          source: edge.source,
          sourceHandle: HANDLE_CONTENT,
          target: edge.target,
          targetHandle: contentTargetHandleForNode(tgt.type),
        })
      }
      if (includeRubric) {
        edges.push({
          id: `${edge.id}-rubric`,
          source: edge.source,
          sourceHandle: HANDLE_RUBRIC,
          target: edge.target,
          targetHandle: rubricTargetHandleForNode(tgt.type),
        })
      }
      continue
    }

    let next = edge
    if (
      tgt.type === 'ai' &&
      (targetHandle === HANDLE_SUBMISSION ||
        targetHandle === HANDLE_CONTENT ||
        targetHandle === HANDLE_RUBRIC ||
        targetHandle === HANDLE_CONTEXT)
    ) {
      next = { ...edge, targetHandle: HANDLE_AI_INPUT }
      changes++
    }
    edges.push(next)
  }

  return {
    graph: { version: graph.version ?? WORKFLOW_VERSION, nodes, edges },
    changes,
  }
}
